types/src/ledger_info.rs

```diff
@@ -373,6 +376,175 @@ impl LedgerInfoWithPartialSignatures {
+        }
+    }
+
+    pub fn commit_info(&self) -> &BlockInfo {
+        self.ledger_info.commit_info()
+    }
+
+    fn add_verified_signature(
+        &mut self,
+        validator: AccountAddress,
+        signature: bls12381::Signature,
+    ) {
+        self.verified_signatures.add_signature(validator, signature);
+        self.unverified_signatures.remove_signature(validator);
+    }
+
+    fn add_unverified_signature(
+        &mut self,
+        validator: AccountAddress,
+        signature: bls12381::Signature,
+    ) {
+        if self.verified_signatures.contains_voter(&validator) {
+            return;
+        }
+        self.unverified_signatures
+            .add_signature(validator, signature);
+    }
+
+    pub fn add_signature(
+        &mut self,
+        validator: AccountAddress,
+        signature: bls12381::Signature,
+        verification_status: VerificationStatus,
+    ) {
+        match verification_status {
+            VerificationStatus::Verified => self.add_verified_signature(validator, signature),
+            VerificationStatus::Unverified => self.add_unverified_signature(validator, signature),
+        };
+    }
+
+    pub fn verified_voters(&self) -> impl Iterator<Item = &AccountAddress> {
+        self.verified_signatures.signatures().keys()
+    }
+
+    pub fn unverified_voters(&self) -> impl Iterator<Item = &AccountAddress> {
+        self.unverified_signatures.signatures().keys()
+    }
+
+    // Collecting all the authors from verified signatures, unverified signatures and the aggregated signature.
+    pub fn all_voters(&self) -> impl Iterator<Item = &AccountAddress> {
+        self.verified_voters().chain(self.unverified_voters())
+    }
+
+    pub fn check_voting_power(
+        &self,
+        verifier: &ValidatorVerifier,
+        check_super_majority: bool,
+    ) -> std::result::Result<u128, VerifyError> {
+        let all_voters = self.all_voters();
+        verifier.check_voting_power(all_voters, check_super_majority)
+    }
+
+    fn try_aggregate(
+        &mut self,
+        epoch_state: &EpochState,
+    ) -> Result<AggregateSignature, VerifyError> {
+        self.check_voting_power(&epoch_state.verifier, true)?;
+
+        let all_signatures = self
+            .verified_signatures
+            .signatures_iter()
+            .chain(self.unverified_signatures.signatures_iter());
+
+        epoch_state.verifier.aggregate_signatures(all_signatures)
+    }
+
+    /// Merge unverified signatures into verified signatures if they are valid.
+    fn merge_signatures(&mut self, verifier: &ValidatorVerifier, need_verify: bool) {
+        let unverified_signatures =
+            mem::replace(&mut self.unverified_signatures, PartialSignatures::empty()).unpack();
+        let valid_signatures: Vec<_> = unverified_signatures
+            .into_par_iter()
+            .flat_map(|(account_address, signature)| {
+                if !need_verify
```

Comment:
The code certainly looks cleaner.
But when need_verify is false, we spawn up the parallel threads in into_par_iter() which is an extra effort even though we don't need them. Is it fine?

Reply:
it's probably fine, I do wonder if we should turn the btree map into a vector and have each task processing multiple signatures though

Reply:
I will implement this recommendation in the next PR to avoid rebasing again.